### PR TITLE
[Review] Support for nodeset xml import with default nodestore

### PR DESCRIPTION
**Pull request by matkonnerth**
This PR plugs in a xml sax parser (from libxml). Nodestore_xml is relying on a valid nodeset, nodes are directly loaded into the nodestore, there are no checks implemented (e.g. check for invalid references).

Example is provided by server_import.c, multiple nodeset can be provided on the commandline (in depending order, e.g. server_nodestore_switch DI PlcOpen).

### CODE REVIEW

File: include/open62541/server.h [hunk 1]
```diff
@@ -85,6 +85,9 @@ UA_Server_run_iterate(UA_Server *server, UA_Boolean waitInternal);
 UA_StatusCode UA_EXPORT
 UA_Server_run_shutdown(UA_Server *server);
 
+void *
+UA_Server_getNsCtx(UA_Server *server);
```
Comment (matkonnerth) [inline include/open62541/server.h]: to use the nodestore interface declared in nodestore.h (for example _UA_Nodestore_newNode(void* nsCtx, UA_NodeClass nodeClass)_ ) the nodestoreContext is needed. Therefore I've added this call in server.h. Do you think there is a better way to do this?

File: plugins/xmlimport/todo.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+support loading of extensions
+loading of values:
+    I think here we would need more infos about a struct?
+rethink char* handling: hashtable to reduce duplicate strings
+loading of datatypes:
+    nodes are loaded, but the I think it don't makes sense to generate the Types_Array.
+    How would you then instantiate for example a struct? There's no c type generated for it?
```
Comment (ccvca) [inline plugins/xmlimport/todo.txt]: When you are programming a generic server application, you can read/write these structs by direc memory operations (set fields at `void * + Offset`) and work inside the application with the names of the members (which can be read from the XML).
But this is definitely an advanced topic, which in my view does not need to be dealt with during the initial implementation.

Comment (matkonnerth) [reply plugins/xmlimport/todo.txt]: Thank you for your comment. Did I understand that right that you mean that loading values (also user defined struct values) should not be part of an initial implementation?

Nevertheless, for loading the value into a variable node like this https://github.com/open62541/open62541/blob/6427010002a44a927dd2ba22b51acf8d31659329/examples/nodeset/testnodeset.xml#L264-L281

the whole information about the struct type will be needed (memSize, padding...). So I would also build up a DataTypeArray, consisting the information about the datatypes in the file, add it to the custom types and then set the variable values with this information available.

Comment (ccvca) [reply plugins/xmlimport/todo.txt]: I'm not a Maintainer, so skipping the loading of structured valued is my personal opinion. If you are already half way done, you can also open an additional pull request with this part.

Parsing structured Datatypes:
When a structured datatype is already known in the Server, you can get all the requrired Information from the UA_TYPES (or the user-Version of it) . Each Type is described by this defintion:
https://github.com/open62541/open62541/blob/6427010002a44a927dd2ba22b51acf8d31659329/include/open62541/types.h#L854-870
By utilizing this structure, you can map the name (from the XML) to size, and padding and all the required informations (if type names are enabled).

**Comment by matkonnerth** [comment]
Think it's time for a review, some points that I'm aware of:
- loading of values is quite new and needs refactoring
- datatypes are loaded, but custom types are not filled at the moment (wip here https://github.com/matkonnerth/open62541/tree/xmlimportWithTypes)

File: plugins/xmlimport/nodesetLoader.c [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            enterUnknownState(pctx);
+            break;
+        case PARSER_STATE_NODE:
+            if(!strcmp(localname, DISPLAYNAME)) {
+                pctx->state = PARSER_STATE_DISPLAYNAME;
+            } else if(!strcmp(localname, REFERENCES)) {
+                pctx->state = PARSER_STATE_REFERENCES;
+            } else if(!strcmp(localname, DESCRIPTION)) {
+                pctx->state = PARSER_STATE_DESCRIPTION;
+            }else if(!strcmp(localname, "Value")) {
```
Comment (ccvca) [inline plugins/xmlimport/nodesetLoader.c]: "Value" is not a define?

File: plugins/xmlimport/value.c [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+setNotImplemented(uintptr_t adr, char *value) {
+    UA_assert(false && "not implemented");
+}
+
+static const ConversionFn conversionTable[UA_DATATYPEKINDS] = {
+    setBoolean,        setSByte,          setByte,           setInt16,
+    setUInt16,         setInt32,          setUInt32,         setInt64,
+    setUInt64,         setFloat,          setDouble,         setString,
+    setDateTime,       setNotImplemented,
+    setString,  // handle like bytestring like string
```
Comment (ccvca) [inline plugins/xmlimport/value.c]: typo in comment, the first 'like' is too much

File: plugins/xmlimport/value.c [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    setNotImplemented, setLocalizedText,  setNotImplemented, setNotImplemented,
+    setNotImplemented, setNotImplemented, setNotImplemented,
+    setInt32,  // handle enum the same way like int32
+    setNotImplemented, setNotImplemented, setNotImplemented, setNotImplemented};
+
+static void
+setScalarValue(Value *val, const UA_DataType *type, size_t padding, char *value) {
+    uintptr_t adr = (uintptr_t)val->value + val->offset + padding;
+    if(value) {
+        conversionTable[type->typeKind](adr, value);
```
Comment (ccvca) [inline plugins/xmlimport/value.c]: Overflow check for type->typeKind missing?

File: tests/nodeset-import/check_nodeset_importSpeed.c [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (ccvca) [inline tests/nodeset-import/check_nodeset_importSpeed.c]: Whats the total time of this test? If it takes a serious amount of time the CI-times might inrease further. (20min for each travis build is already quite some time at the moment)

File: plugins/xmlimport/conversion.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
+#ifndef CONVERSION_H
+#define CONVERSION_H
+#include <open62541/types.h>
+
+static inline UA_Boolean
+isNodeId(const char *s) {
+    if(!strncmp(s, "ns=", 3) || !strncmp(s, "i=", 2) || !strncmp(s, "s=", 2)) {
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/conversion.h]: done

File: plugins/xmlimport/conversion.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    if(strcmp(s, "true")) {
+        return UA_FALSE;
+    }
+    return UA_TRUE;
+}
+
+// todo: handle guid, bytestring
+static inline UA_NodeId
+extractNodeId(char *s) {
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/conversion.h]: added const specifier

File: plugins/xmlimport/nodeset.c [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+#define ATTRIBUTE_SYMMETRIC "Symmetric"
+#define ATTRIBUTE_ALIAS "Alias"
+
+typedef struct {
+    const char *name;
+    char *defaultValue;
+    bool optional;
+} NodeAttribute;
+
+NodeAttribute attrNodeId = {ATTRIBUTE_NODEID, NULL, false};
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: done

File: plugins/xmlimport/nodeset.c [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+    UA_NODECLASS_VARIABLETYPE};
+
+struct Alias {
+    char *name;
+    UA_NodeId id;
+};
+
+struct TNamespace {
+    UA_UInt16 idx;
+    char *name;
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: done

File: plugins/xmlimport/nodeset.c [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+    TNamespace *ns;
+    addNamespaceCb cb;
+} TNamespaceTable;
+
+typedef struct {
+    UA_NodeId *src;
+    UA_NodeReferenceKind *ref;
+} TRef;
+
+struct MemoryPool;
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: done

File: plugins/xmlimport/nodeset.c [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+                getAttributeValue(nodeset, &attrIsAbstract, attributes, attributeSize));
+            break;
+        case UA_NODECLASS_METHOD:
+            //((UA_MethodNode *)node)->executable =
+            //    isTrue(extractNodedId(namespaces, getAttributeValue(&attrParentNodeId,
+            //    attributes, attributeSize));
+            break;
+        case UA_NODECLASS_REFERENCETYPE:
+            break;
+        default:;
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: You are right , this one is missing.

File: plugins/xmlimport/nodeset.c [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+    UA_NodeReferenceKind *newRef = refs + node->referencesSize - 1;
+    newRef->referenceTypeId = refTypeId;
+    newRef->isInverse = !isForward;
+    newRef->targetIdsSize = 0;
+    newRef->targetIds = NULL;
+    return newRef;
+}
+
+static bool
+isBidirectionalReference(Nodeset *nodeset, const UA_NodeId *refId) {
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: This array contains the ReferenceTypes, for which automatically the inverse reference is added (also when only one direction of the reference is modelled in the xml). This is necessary for hierachical refs, and I think at least for hasEncoding and subtypes of it.

I think I could get all hierachical references from the server, also hasEncoding and subtypes and get rid of that array.

File: plugins/xmlimport/nodeset.h [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+typedef enum {
+    NODECLASS_OBJECT = 0,
+    NODECLASS_OBJECTTYPE = 1,
+    NODECLASS_VARIABLE = 2,
+    NODECLASS_DATATYPE = 3,
+    NODECLASS_METHOD = 4,
+    NODECLASS_REFERENCETYPE = 5,
+    NODECLASS_VARIABLETYPE = 6
+} TNodeClass;
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.h]: done

File: plugins/xmlimport/nodesetLoader.c [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+    UA_NodeReferenceKind *refKind;
+    Alias *alias;
+    void *userContext;
+    char *onCharacters;
+    Nodeset *nodeset;
+    struct Value *val;
+};
+
+struct TParserCtx;
+typedef struct TParserCtx TParserCtx;
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodesetLoader.c]: done

File: plugins/xmlimport/nodesetLoader.c (hunk 3 above)
Comment (matkonnerth) [reply plugins/xmlimport/nodesetLoader.c]: done

File: plugins/xmlimport/value.c (hunk 4 above)
Comment (matkonnerth) [reply plugins/xmlimport/value.c]: done

File: tests/nodeset-import/check_nodeset_importSpeed.c (hunk 6 above)
Comment (matkonnerth) [reply tests/nodeset-import/check_nodeset_importSpeed.c]: should be below 10-20s

File: plugins/xmlimport/nodeset.c [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+    size_t charsSize;
+    TNamespaceTable *namespaceTable;
+    size_t hierachicalRefsSize;
+    UA_NodeId *hierachicalRefs;
+    struct MemoryPool *refPool;
+    UA_Server *server;
+};
+
+/* start with hierachical and hasEncoding refs */
+UA_NodeId bidirectionalRefs[MAX_BIDIRECTIONAL_REFS] = {
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: cannot make it const, at the moment all hierachical refs and subtypes of them are added to this array during parsing. But should rethink (see comment below, I will close this.

**Comment by matkonnerth** [comment]
thanks for your comments @ccvca . You are welcome if you want so solve some of them, or generally work on this pr.

File: plugins/xmlimport/nodeset.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: done

File: plugins/xmlimport/value.c (hunk 5 above)
Comment (matkonnerth) [reply plugins/xmlimport/value.c]: think not necessary here, typekind should be set to one of the enum values in types.h.

File: plugins/xmlimport/nodesetLoader.c [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+#include <open62541/plugin/nodestore.h>
+
+#include <stdio.h>
+#include <string.h>
+
+#include "nodeset.h"
+#include "value.h"
+#include <libxml/SAX.h>
+
+#define OBJECT "UAObject"
```
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodesetLoader.c]: will not improve the code, have to manually add the attributes of a node.

e.g. UA_ObjectTypeNode->isAbstract = getAttributeValue(&attrIsAbstract) is needed, don't know how to do this in a generic way.

File: plugins/xmlimport/nodeset.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (matkonnerth) [reply plugins/xmlimport/nodeset.c]: won't fix this at the moment, if there's no interest in this PR

File: include/open62541/server.h (hunk 1 above)
Comment (matkonnerth) [reply include/open62541/server.h]: UA_Nodestore* now available from serverConfig, so I removed this hack.

**Comment by MartinaFelberFEM** [comment]
Hi @matkonnerth !

I would like to contribute to your pull request.
Unfortunately I am not able to fork your repository matkonnerth/open62541 and make a pull request. It seems that this is not supported because I have a fork of open62541 too. 
I am not sure how to work around this yet.

I have just had a look through your code.
I think I have found an issue with the namespace indices of BrowseNames. 
I have started a review, don't know if a comment would have been better? I am not so familiar with GitHub.

I would have liked to fix it, but without a fork of your repository I don't know how I can contribute.
I will try to get more familiar with your work and I would very much appreciate it, if you could give me some suggestions how I can contribute.
E.g. todo's, bugs, refactoring, tests, etc.

Thanks

**Comment by matkonnerth** [comment]
Hi @MartinaFelberFEM !
I opened this pull request a time ago when adding nodes via _UA_Server_addVariableNode_ (and friends) was quite slow. It took about 30s-60s on an i7 with 100k nodes. So i use in this pull request the interface of the nodestore. That has the advantages that it is really fast to add a nodeset and don't have to sort the nodes before adding it. The downside is, that it doesn't do any check if the node you are adding is valid (valid datatypes, valid references and so on). Another disadvantage is, that it uses a interface, that is not usable from open62541. For example, if the UA_Node struct is changed (the internal represenation of a node, declared in nodestore.h), you have to touch this importer and fix it.

Another option is shown here: https://github.com/matkonnerth/nodesetLoader
It also uses libxml and a sax parser like in this PR but is independent of the datatypes of open62541. It also does the sorting. For adding the parsed nodes to the server (to open62541, but could be also another one) I've done here a conversion: https://github.com/matkonnerth/ModernOPC/blob/master/wrapper/src/import/import.cpp#L126

Short story: 
Option 1: loading nodes direct via nodestore interface (this PR)
Option 2: loading nodes via the official interface (nodesetLoader)

I'm using currently the second option because I don't see the big difference in loading time. I have to verify this again, but it option 2 takes about 1-2s to load the 100k nodes nodesets on a i7.

So the first thing to do would be here to think of which direction we want to go, maybe together with @jpfr .

We can also talk about this the upcoming week and I can give a detailled overview of the two solutions.

**Regarding the issues**
you can make a review on this PR. We cannot have tests enough, best thing would be a test case that fails and then we fix it. 
**Regarding testing**
The test case  very useful and don't have to be rewritten, no matter if we choose option 1 or 2 when the are written in a way like:
load a small test nodeset
check with e.g. a read/browse if the nodeset is setup as expected.

Regarding the "fork of a fork": I think that's not supported, if you have already forked the open62541 repo (no matter why github is not supporting this).

**Comment by Rondine89** [comment]
Hi!
I don't know if this is the right place to ask, if not, just delete this post.
I'm having trouble with the https://github.com/matkonnerth/open62541/blob/xmlimportWithTypes/plugins/xmlimport/nodesetLoader.c file, because of the sax.h. I've include in the code the libxml header files but, when i compile, I get errors in the static int read_xmlfile(FILE *f, TParserCtx *parserCtxt) functions and these errors are related to the API of libxml.
I've tried to include the libxml2.a in the library path, but that leads to other errors.

How did you managed the libxml2 when you developed this program?

**Comment by matkonnerth** [comment]
Hi @Rondine89 , I've installed libxml2 on my machine and normally cmake finds this library.

**Comment by matkonnerth** [comment]
I will close this PR. Adding nodes via the add_Nodes interfaces has been improved, therefore sacrifizing checks for import speed is not necessary anymore.

A lib for importing nodesets via the addNodes service kann be found here, please note that's not production ready, but will be improved:
https://github.com/matkonnerth/nodesetLoader

**Pull request closed without merging**